### PR TITLE
feat(amqp_invoker): asynchronous consumer-producer

### PR DESCRIPTION
**Pull request by ahuman-bean**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,4 +4,7 @@ falcon
 ansicolors
 click-default-group
 pika
-PyYAML
\ No newline at end of file
+PyYAML
+aio-pika
+retry
+aiomisc
```
Comment (zekenautilus) [inline requirements.txt:10]: are all of these packages cool w the versions of python we're trying to support? 3.6 -> 3.9

File: src/amqp_invoker.py [hunk 2]
```diff
@@ -20,51 +29,142 @@ def set_param(host: str, param_key: str, param_val: str) -> str:
...
-        queue_name = self._invocable.config.func
-        queue_name_error = f'{queue_name}_error'
-        exchange_name = self._invocable.config.exchange
-        channel.queue_declare(queue=queue_name)
-        channel.queue_declare(queue=queue_name_error)
-        channel.exchange_declare(exchange_name, exchange_type='topic', passive=False, durable=True, auto_delete=False, internal=False, arguments=None)
-        channel.queue_bind(exchange=exchange_name, queue=queue_name, routing_key=str(self._invocable.config.subtopic))
-        return channel, queue_name, queue_name_error
+
+        self.url = set_param(host, 'heartbeat', str(heartbeat)) if heartbeat else host
```
[the remark this answers is not in the document]
Comment (ahuman-bean) [reply src/amqp_invoker.py:38]: we still need the heartbeat to maintain the connection. w/ the broker.

async just means that we are using promises/futures to express asynchronous behavior synchronously (procedurally)

File: src/amqp_invoker.py [hunk 3]
```diff
@@ -20,51 +29,151 @@ def set_param(host: str, param_key: str, param_val: str) -> str:
...
+                    futures = [
+                        self.consume(channel, queue, self.on_message),
+                        # TODO(ahuman-bean): implement error handling
+                        # self.consume(channel, queue_error, self.on_error)
+                    ]
+
+                    # Gathers all workers as futures and runs them concurrently on the underlying execution context
+                    await asyncio.gather(*futures)
+                except aio_pika.exceptions.ConnectionClosed:
+                    pass
```
[the remark this answers is not in the document]
Comment (zekenautilus) [reply src/amqp_invoker.py:104]: or is this part of what the decorator handles?

File: src/amqp_invoker.py [hunk 4]
```diff
@@ -20,51 +29,151 @@ def set_param(host: str, param_key: str, param_val: str) -> str:
...
-                                          routing_key=str(self._invocable.config.pubtopic), body=json.dumps(payload))
+                # `self.do_work` is guaranteed to run on a separate thread, which is recommended for potentially long-running background tasks
+                async with self.do_work(data_in) as generator:
+                    async for data_out in generator:
+                        await exchange.publish(message=aio_pika.Message(body=json.dumps(data_out).encode()), routing_key=str(self._invocable.config.pubtopic))
+
             except Exception as err:  # pylint: disable=broad-except
                 data_in['error'] = str(err)
-                channel.basic_publish(exchange='', routing_key=queue_name_error, body=json.dumps(data_in))
+                # TODO(ahuman-bean): cleaner error messages
```
Comment (zekenautilus) [inline src/amqp_invoker.py:149]: https://github.com/mattian7741/ergo/issues/23

File: src/amqp_invoker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahuman-bean) [reply src/amqp_invoker.py:104]: yeah I think this can be removed, this was prior to me adding the decorator and messing around w/ `RobustConnection`

**Pull request closed without merging**